src/mouse_events.cpp

```diff
@@ -1405,19 +1406,17 @@ std::set<map_location> mouse_handler::get_adj_enemies(const map_location& loc, i
 void mouse_handler::show_attack_options(const unit_map::const_iterator& u)
 {
 	// Cannot attack if no attacks are left.
-	if(u->attacks_left() == 0) {
+	if(u->attacks_left() == 0 || !u->has_attacks()) {
 		return;
 	}
 
 	// Get the teams involved.
 	const team& cur_team = current_team();
 	const team& u_team = pc_.gamestate().board_.teams_[u->side() - 1];
 
-	// Check each adjacent hex.
-	adjacent_loc_array_t adj;
-	get_adjacent_tiles(u->get_location(), adj.data());
-
-	for(const map_location& loc : adj) {
+	// Check each tile in the unit's attack range
+	auto range = get_tiles_in_range(u->get_location(), u->min_attack_range(), u->max_attack_range());
```

Comment:
you added an assertion on about hte parameters on get_tiles_in_range but i see no code here that makes sure the assertion is true.

Reply:
furthermore it might happen that your units has multiple attacks, one for long range and one for short range, and on first sight this looks like it will make all available.

Reply:
> but i see no code here that makes sure the assertion is true.

Nevermind it seems to be slot in the setter in attack min/max range

Reply:
It's in attack_type.cpp. min_range is always initialized to at least 1 and max_range to at least min_range.

However, I now realize that later on the >0 constraint is not enforced when modifying the values. I'll fix that.

Reply:
> furthermore it might happen that your units has multiple attacks, one for long range and one for short range, and on first sight this looks like it will make all available.

This part of the code only controls which units can be targeted. So any enemy that you can reach with at least one weapon should be highlighted. Later, the attack dialog will filter out the weapons without proper range (this was already there).